Change API and framework Active Response message format to JSON

mcarmona99 opened the pull request:
|Related issue|
|---|
| #7210 |

This pull request closes #7210.

In this PR, I have changed the Active Response messages to JSON when needed.

The message sent in `/cluster/configuration/validation`, `/manager/configuration/validation`, `/cluster/restart` and `/manager/restart` has been changed to the standard JSON format, as detailed in the issue.

The message sent in `/agents/restart`, /`agents/{agent_id}/restart` and `/agents/group/{group_id}/restart` changes to JSON format when the agent version is `>= 4.2.0`.

The message sent in `/active-response` will be the same if the agent version is `< 4.2.0` and it will be a JSON message when agent version is `>=4.2.0`. The message is built using the JSON body when using the API. I have included a new field in the JSON body: `alert`.

I have updated the active-response integration tests (rbac and no rbac) with the new exception added and I have sent the command `restart-wazuh0` instead of `restart-ossec0`. This change has been also added to the `ar` configurations (`ar.conf` and `mergedx.mg`).

The following unit tests have been updated too:

	pytest wazuh/core/tests/test_agent.py -> added versions and test if the message sent is the correct one
		======================================== 153 passed, 11 warnings in 1.05s =========================================

	pytest wazuh/core/tests/test_configuration.py -> added ar command restart-wazuh0 for scripts
		=============================================== 34 passed in 0.16s ================================================

	pytest wazuh/core/tests/test_active_response.py -> added test case for create_json_message function
		=============================================== 17 passed in 0.09s ================================================

	pytest  framework/wazuh/tests/test_active_response.py -> update test for run_command
		=============================================== 12 passed in 0.10s ===============================================


### Test results:

```
test_active_response_endpoints.tavern.yaml 
	 2 passed, 4 warnings

test_rbac_black_active_response_endpoints.tavern.yaml 
	 2 passed, 4 warnings

test_rbac_white_active_response_endpoints.tavern.yaml 
	 2 passed, 4 warnings
```

```
test_cluster_endpoints.tavern.yaml 
	 52 passed, 1 xpassed, 70 warnings
	 
test_manager_endpoints.tavern.yaml 
	 33 passed, 36 warnings
	 
test_agent_PUT_endpoints.tavern.yaml 
	 9 passed, 11 warnings
```

```
pytest wazuh/core/cluster/tests/
================================ 55 passed, 2 warnings in 0.40s =================================

pytest wazuh/core/cluster/dapi/tests/
================================ 22 passed, 11 warnings in 0.54s ================================
```

Regards,
Manuel.

davidjiglesias commented on CHANGELOG.md:
```diff
@@ -11,7 +11,11 @@ All notable changes to this project will be documented in this file.
 
 - **API:**
   - Added new endpoint to get agent stats from different components. ([#7128](https://github.com/wazuh/wazuh/issues/7128))
+  - Change Active Response endpoint function to create messages with new JSON format. ([#7312](https://github.com/wazuh/wazuh/pull/7312))
```
Use past tense. Move to appropriate tag (Changed, not Added). Both entries (API and framework) are too similar, if there is no justification for the framework entry changes, remove it.

mcarmona99 replied:
corrected, sentence moved to `Changed` and removed the framework change entry

davidjiglesias commented on api/api/models/active_response_model.py:
```diff
@@ -10,75 +10,113 @@
+        custom : bool
+            Whether the specified command is a custom command or not
+        arguments : List[str]
+            Command arguments
```
Missing alert

mcarmona99 replied:
corrected

davidjiglesias commented on framework/wazuh/core/active_response.py:
```diff
@@ -61,10 +107,16 @@ def shell_escape(command):
+    msg_queue : str
+        Message previously created, contains what is necessary to launch the active response command in the agent
+    agent_id : str
+        ID specifying the agent the msg_queue will be sent to
```
```suggestion
        ID specifying the agent where the msg_queue will be sent to
```

davidjiglesias commented on framework/wazuh/core/manager.py:
```diff
@@ -273,8 +273,8 @@ def validate_ossec_conf():
         # Message for checking Wazuh configuration
-        execq_msg = 'check-manager-configuration '
-
+        execq_msg = "{\"version\": 1, \"origin\": {\"module\": \"api\"}, \"command\": " \
```
Same as above

mcarmona99 replied:
used json.dumps

davidjiglesias commented on framework/wazuh/core/tests/test_active_response.py:
```diff
@@ -28,6 +27,14 @@ def agent_info(expected_exception):
 
 
+def agent_info_exception_and_version(expected_exception, version):
+    """Returns dict to cause or not a exception code 1651 on active_response.send_command()."""
```
```suggestion
    """Return dict to cause or not a exception code 1651 on active_response.send_command()."""
```

davidjiglesias commented on framework/wazuh/core/tests/test_active_response.py:
```diff
@@ -76,6 +83,44 @@ def test_create_message(expected_exception, command, arguments, custom):
+])
+@patch('wazuh.core.common.ossec_path', new=test_data_path)
+def test_create_json_message(expected_exception, command, arguments, alert):
+    """Check if the json message returned is correct
```
```suggestion
    """Check if the returned json message is correct.
```

davidjiglesias commented on framework/wazuh/core/tests/test_active_response.py:
```diff
@@ -76,6 +83,44 @@ def test_create_message(expected_exception, command, arguments, custom):
+def test_create_json_message(expected_exception, command, arguments, alert):
+    """Check if the json message returned is correct
+
+    Checks if json message returned by create_json_message(...) contains the
```
```suggestion
    Checks if the json message returned by create_json_message(...) contains the
```

davidjiglesias commented on framework/wazuh/core/tests/test_agent.py:
```diff
@@ -1461,11 +1466,26 @@ def test_calculate_status(last_keep_alive, pending, expected_status):
-    send_restart_command('001')
-    mock_ossec_queue.return_value.send_msg_to_agent.assert_called_once_with(ANY, '001')
+def test_send_restart_command(mock_ossec_queue, agents_list, versions_list):
+    """Test that restart_command calls send_msg_to_agent with correct params
```
```suggestion
    """Test that restart_command calls send_msg_to_agent with correct params.
```

mcarmona99 commented on api/test/integration/test_active_response_endpoints.tavern.yaml:
```diff
@@ -6,7 +6,7 @@ marks:
 stages:
 
-  - name: Runs an Active Response command on a specified agent
+  - name: Runs an Active Response command on a specified 4.2.0 agent
```
corrected

mcarmona99 commented on framework/wazuh/active_response.py:
```diff
@@ -3,36 +3,55 @@
+    arguments : list
+        Command arguments.
+    alert : dict
+        Same alert from alerts.json file in JSON format.
```
improved

mcarmona99 commented on framework/wazuh/core/cluster/utils.py:
```diff
@@ -153,8 +153,9 @@ def manager_restart():
-        # msg for restarting Wazuh manager
-        msg = 'restart-wazuh '
+        # json msg for restarting Wazuh manager
+        msg = "{\"version\": 1, \"origin\": {\"module\": \"api/framework\"}, \"command\": \"restart-wazuh\"," \
```
used json.dumps

mcarmona99 commented on api/test/integration/test_active_response_endpoints.tavern.yaml:
```diff
@@ -208,28 +212,42 @@ stages:
+            - error:
+                code: 1652
+              id:
+                - '000'
```
ar module refactored

mcarmona99 commented on api/test/integration/test_active_response_endpoints.tavern.yaml:
```diff
@@ -208,28 +212,42 @@ stages:
                 - '003'
-          total_affected_items: 8
-          total_failed_items: 5
+          total_affected_items: 7
```
ar module refactored

mcarmona99 commented on api/test/integration/test_rbac_black_active_response_endpoints.tavern.yaml:
```diff
@@ -144,19 +144,26 @@ stages:
                 code: 1651
               id:
                 - '010'
+            - error:
```
ar module refactored

mcarmona99 commented on api/test/integration/test_rbac_white_active_response_endpoints.tavern.yaml:
```diff
@@ -146,19 +146,22 @@ stages:
                 code: 1651
               id:
                 - '009'
+            - error:
```
ar module refactored

mcarmona99 commented on framework/wazuh/active_response.py:
```diff
@@ -3,36 +3,55 @@
         except WazuhException as e:
             result.add_failed_item(id_=agent_id, error=e)
+        except KeyError:
+            result.add_failed_item(id_=agent_id, error=WazuhError(1657))
```
ar module refactored

mcarmona99 commented on framework/wazuh/core/agent.py:
```diff
@@ -1131,7 +1131,13 @@ def send_restart_command(agent_id):
+    if WazuhVersion(agent_version) >= WazuhVersion('Wazuh v4.2.0'):
+        ret_msg = oq.send_msg_to_agent(OssecQueue.RESTART_AGENTS_JSON, agent_id)
+    else:
+        ret_msg = oq.send_msg_to_agent(OssecQueue.RESTART_AGENTS, agent_id)
```
ar module refactored

mcarmona99 commented on framework/wazuh/core/exception.py:
```diff
@@ -211,6 +211,8 @@ class WazuhException(Exception):
                'remediation': 'Please, visit the official documentation '
                               f'(https://documentation.wazuh.com/{WAZUH_VERSION}/user-manual/api/reference.html#tag/active-response) '
                               'to get more information about `active-response` API call'},
+        1657: {'message': 'Active response - Cannot send AR command, agent version is null',
```
ar module refactored, exception deleted

mcarmona99 commented on framework/wazuh/core/ossec_queue.py:
```diff
@@ -15,6 +15,8 @@ class OssecQueue:
     # Messages
     HC_SK_RESTART = "syscheck restart"  # syscheck restart
     RESTART_AGENTS = "restart-ossec0"  # Agents, not manager (000)
+    RESTART_AGENTS_JSON = "{\"version\": 1,\"origin\": {\"module\": \"api/framework\"}, \"command\": \"restart-wazuh0\", " \
```
changed to json.dumps, import json too

mcarmona99 commented:
Refactor done to the active response module.

I have change a lot of unit tests and api integration tests due to the fact that a lot of core and SDK functions have changed.

**Test results:**

framework/wazuh/tests/test_active_response.py
========================= 12 passed in 0.10s =========================

framework/wazuh/core/tests/test_configuration.py
========================= 34 passed in 0.15s =========================

framework/wazuh/core/tests/test_agent.py
================== 153 passed, 11 warnings in 1.04s ==================

framework/wazuh/core/tests/test_active_response.py
========================= 14 passed in 0.08s =========================

Regards,
Manuel.

mcarmona99 commented:
Big changes to the AR response module.

I have used the `create_wazuh_socket_message` function to create AR messages sent to `execq` and `arq`.

I have updated the docstrings with the new standard.

Function parameters now include their type and default value. Functions are also including the returned type.

I have also changed hardcoded checks, variables `AR_LEGACY_VERSION`, `RESTART_WAZUH_COMMAND`, etc have been added.



### **Test results:**

```
framework/wazuh/core/tests/test_active_response.py
============================== 14 passed in 0.08s ==============================

framework/wazuh/tests/test_active_response.py
============================== 12 passed in 0.11s ==============================

framework/wazuh/core/tests/test_configuration.py
============================== 34 passed in 0.14s ==============================                        

framework/wazuh/core/tests/test_agent.py 
======================= 153 passed, 11 warnings in 0.93s =======================
```

mcarmona99 commented:
### **API integration tests results:**

```
test_active_response_endpoints.tavern.yaml 
	 2 passed, 4 warnings

test_rbac_black_active_response_endpoints.tavern.yaml 
	 2 passed, 4 warnings

test_rbac_white_active_response_endpoints.tavern.yaml 
	 2 passed, 4 warnings
```

davidjiglesias commented on framework/wazuh/core/tests/test_agent.py:
```diff
@@ -561,21 +564,21 @@ def test_agent_get_key_ko(socket_mock, send_mock):
 
         # Assert methods are called with correct params
-        mock_config.assert_called_once_with('com', 'active-response')
+        mock_config.assert_called_once_with('com', 'active-response', 'Wazuh v3.9.0')
```
I see Wazuh vX.Y.Z in some places and just vX.Y.Z, please unify, i would just use vX.Y.Z

mcarmona99 replied:
Unification: I have used Wazuh vX.Y.Z in all places.

davidjiglesias commented on framework/wazuh/core/tests/test_agent.py:
```diff
@@ -1408,15 +1413,21 @@ def test_agent_getconfig(socket_mock, send_mock, mock_ossec_socket):
+        agent.getconfig('com', None, 'v4.0.0')
+        agent.getconfig(None, 'active-response', 'v4.0.0')
 
     # Agent Wazuh version is lower than v3.7.0
```
Refer to the variable instead of v3.7.0

davidjiglesias commented on framework/wazuh/core/tests/test_agent.py:
```diff
@@ -1461,11 +1472,26 @@ def test_calculate_status(last_keep_alive, pending, expected_status):
+        for agent_id, agent_version in zip(agents_list, versions_list):
+            send_restart_command(agent_id, agent_version['version'])
+            expected_msg = mock_ossec_queue.RESTART_AGENTS_JSON if WazuhVersion(
+                agent_version['version']) >= WazuhVersion('Wazuh v4.2.0') else mock_ossec_queue.RESTART_AGENTS
```
Use variable for version

mcarmona99 commented on framework/wazuh/active_response.py:
```diff
@@ -3,36 +3,62 @@
-@expose_resources(actions=['active-response:command'], resources=['agent:id:{agent_list}'])
-def run_command(agent_list=None, command=None, arguments=None, custom=False):
-    """Run AR command in a specific agent
+    Raises
```
Removed raises from docstring

mcarmona99 commented on framework/wazuh/core/active_response.py:
```diff
@@ -30,46 +49,124 @@ def create_message(command, custom, arguments):
+        raise WazuhError(1750)
+
+    # Create classic msg or JSON msg depending on the agent version
+    if WazuhVersion(agent_version) >= WazuhVersion('Wazuh v4.2.0'):
```
The variable was created, I didn't change the version to the variable there.

mcarmona99 commented on framework/wazuh/core/agent.py:
```diff
@@ -1059,18 +1069,31 @@ def unset_single_group_agent(agent_id, group_id, force=False):
+        dict
+            Agent's active configuration.
+        """
+        required_version = 'v3.7.0'
```
Added to common
